ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/CreateBcache/CreateBcache.tsx

```diff
@@ -0,0 +1,143 @@
+import { useDispatch, useSelector } from "react-redux";
+import * as Yup from "yup";
+
+import CreateBcacheFields from "./CreateBcacheFields";
+
+import FormCardButtons from "app/base/components/FormCardButtons";
+import FormikForm from "app/base/components/FormikForm";
+import { useMachineDetailsForm } from "app/machines/hooks";
+import { actions as machineActions } from "app/store/machine";
+import machineSelectors from "app/store/machine/selectors";
+import type { Disk, Machine, Partition } from "app/store/machine/types";
+import { BcacheModes } from "app/store/machine/types";
+import { isBcache, isCacheSet, isDisk } from "app/store/machine/utils";
+import type { RootState } from "app/store/root/types";
+
+export type CreateBcacheValues = {
+  cacheMode: BcacheModes;
+  cacheSetId: string;
+  fstype?: string;
+  mountOptions?: string;
+  mountPoint?: string;
+  name: string;
+  tags: string[];
+};
+
+type Props = {
+  closeExpanded: () => void;
+  storageDevice: Disk | Partition;
+  systemId: Machine["system_id"];
+};
+
+const CreateBcacheSchema = Yup.object().shape({
+  cacheMode: Yup.string().required("Cache mode is required"),
+  cacheSetId: Yup.number().required("Cache set is required"),
+  fstype: Yup.string(),
+  mountOptions: Yup.string(),
+  mountPoint: Yup.string().when("fstype", {
+    is: (val: CreateBcacheValues["fstype"]) => Boolean(val),
+    then: Yup.string().matches(/^\//, "Mount point must start with /"),
+  }),
+  name: Yup.string().required("Name is required"),
+  tags: Yup.array().of(Yup.string()),
+});
+
+const getInitialName = (disks: Disk[]) => {
+  if (!disks || disks.length === 0) {
+    return "bcache0";
+  }
+  const bcacheCount = disks.reduce<number>(
+    (count, disk) => (isBcache(disk) ? count + 1 : count),
+    0
+  );
+  return `bcache${bcacheCount}`;
+};
+
+export const CreateBcache = ({
+  closeExpanded,
+  storageDevice,
+  systemId,
+}: Props): JSX.Element | null => {
+  const dispatch = useDispatch();
+  const { errors, saved, saving } = useMachineDetailsForm(
+    systemId,
+    "creatingBcache",
+    "createBcache",
+    () => closeExpanded()
+  );
+  const machine = useSelector((state: RootState) =>
+    machineSelectors.getById(state, systemId)
+  );
+
+  if (machine && "disks" in machine) {
+    const cacheSets = machine.disks.filter((disk) => isCacheSet(disk));
+
+    if (cacheSets.length === 0) {
+      closeExpanded();
```

Comment:
Can this case ever happen? It seems like we shouldn't be able to do the action that opens this form if it will immediately close...

Reply:
I was thinking it could be possible if someone else deleted the last remaining cache set while you had the bcache form open. It's a pretty unlikely scenario... but that's why it's there. :man_shrugging:

Reply:
OK I see, might be worth adding a comment for next time I stare blankly at this bit of code :).

Reply:
Will do!